### PR TITLE
feat(gitlabci): add support for multidoc yaml

### LINKED ISSUE #28416: gitlabci manager: support multidoc yaml
### Discussed in https://github.com/renovatebot/renovate/discussions/28386

<div type='discussions-op-text'>

<sup>Originally posted by **benedikt-bartscher** April 13, 2024</sup>
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gitlab - Renovate 37.291.0

### Please tell us more about your question or problem

With gitlabs new CI Components gitlab-ci files can be multidocument yaml ([docs](https://docs.gitlab.com/ee/ci/yaml/inputs.html#define-input-parameters-with-specinputs)). Renovate currently fails parsing multidocument gitlab-ci files.  

Something like https://github.com/renovatebot/renovate/pull/8828/files is needed

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: YAML exception extracting GitLab CI includes (repository=senbax/gitlabci/docker, packageFile=templates/docker.yml)
       "err": {
         "name": "YAMLException",
         "reason": "expected a single document in the stream, but found more",
         "message": "expected a single document in the stream, but found more",
         "stack": "YAMLException: expected a single document in the stream, but found more\n    at load (/usr/local/renovate/node_modules/.pnpm/js-yaml@4.1.0/node_modules/js-yaml/lib/loader.js:1722:9)\n    at parseSingleYaml (/usr/local/renovate/lib/util/yaml.ts:65:29)\n    at Object.extractPackageFile (/usr/local/renovate/lib/modules/manager/gitlabci-include/extract.ts:75:32)\n    at extractPackageFile (/usr/local/renovate/lib/modules/manager/index.ts:75:9)\n    at getManagerPackageFiles (/usr/local/renovate/lib/workers/repository/extract/manager-files.ts:45:43)\n    at /usr/local/renovate/lib/workers/repository/extract/index.ts:57:28\n    at async Promise.all (index 1)\n    at extractAllDependencies (/usr/local/renovate/lib/workers/repository/extract/index.ts:54:26)\n    at extract (/usr/local/renovate/lib/workers/repository/process/extract-update.ts:140:28)\n    at extractDependencies (/usr/local/renovate/lib/workers/repository/process/index.ts:154:26)\n    at Object.renovateRepository (/usr/local/renovate/lib/workers/repository/index.ts:69:9)\n    at attributes.repository (/usr/local/renovate/lib/workers/global/index.ts:217:11)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:202:7)\n    at /usr/local/renovate/lib/renovate.ts:18:22"
       }

```

</details>
</div>

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Use `parseYaml` instead of `parseSingleYaml`
- Add test

## Context
Closes: https://github.com/renovatebot/renovate/issues/28416

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gitlabci/extract.spec.ts [hunk 1]
```diff
@@ -38,6 +38,14 @@ describe('modules/manager/gitlabci/extract', () => {
       ).toBeNull();
     });
 
+    it('extracts from multidoc yaml', async () => {
+      const res = await extractAllPackageFiles(config, [
+        'lib/modules/manager/gitlabci/__fixtures__/gitlab-ci.multi-doc.yaml',
+      ]);
+      expect(res).toMatchSnapshot();
```
Comment (secustor) [inline lib/modules/manager/gitlabci/extract.spec.ts:45]: No new snapshots, please.

File: lib/modules/manager/gitlabci/extract.ts [hunk 2]
```diff
@@ -165,43 +165,54 @@ export function extractPackageFile(
...
-        switch (property) {
-          case 'image':
-            {
-              const dep = extractFromImage(
-                value as Image,
-                config.registryAliases,
-              );
-              if (dep) {
-                deps.push(dep);
+    const docs = parseYaml<GitlabPipeline>(
```
Comment (secustor) [inline lib/modules/manager/gitlabci/extract.ts:168]: As we use types here, do you have considered using schemas?

Comment (RahulGautamSingh) [reply lib/modules/manager/gitlabci/extract.ts:168]: There is already issue for that, so I thought do do that in another PR.

File: lib/modules/manager/gitlabci/extract.spec.ts [hunk 3]
```diff
@@ -38,6 +38,21 @@ describe('modules/manager/gitlabci/extract', () => {
...
+        'lib/modules/manager/gitlabci/__fixtures__/gitlab-ci.multi-doc.yaml',
+      ]);
+      expect(res).toHaveLength(3);
+
+      const deps: PackageDependency[] = [];
+      res?.forEach((e) => {
+        e.deps.forEach((d) => {
+          deps.push(d);
+        });
+      });
```
Comment (secustor) [inline lib/modules/manager/gitlabci/extract.spec.ts:52]:
```suggestion
      const deps: PackageDependency[] = res.map(entry => e.deps).flat()
```

File: lib/modules/manager/gitlabci/extract.ts (hunk 2 above)
Comment (secustor) [reply lib/modules/manager/gitlabci/extract.ts:168]: No it is fine for now.

**Review by secustor — APPROVED** [review]

**Pull request merged**